File: CS/Bsv.cs
#nullable enable

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;

static byte[] EncodeBsv(string?[][] jaggedArray) {
	var parts = new List<byte[]>();
	var lineBreakByte = new byte[]{0xFF};
	var valueSeparatorByte = new byte[]{0xFE};
	var nullValueByte = new byte[]{0xFD};
	var emptyStringByte = new byte[]{0xFC};
	var encoder = new UTF8Encoding(false, true);
	var isFirstLine = true;
	foreach (var line in jaggedArray) {
		if (isFirstLine == false) { parts.Add(lineBreakByte); }
		isFirstLine = false;
		var isFirstValue = true;
		foreach (var value in line) {
			if (isFirstValue == false) { parts.Add(valueSeparatorByte); }
			isFirstValue = false;
			if (value == null) { parts.Add(nullValueByte); }
			else if (value.Length == 0) { parts.Add(emptyStringByte); }
			else { parts.Add(encoder.GetBytes(value)); }
		}
	}
	var result = new byte[parts.Sum(part => part.Length)];
	var stream = new MemoryStream(result);
    foreach (var part in parts) { stream.Write(part, 0, part.Length); }
	return result;
}

static string?[][] DecodeBsv(byte[] bytes) {
	var decoder = new UTF8Encoding(false, true);
	var result = new List<string?[]>();
	var currentLine = new List<string?>();
	var currentIndex = -1;
	for (;;) {
		var lastIndex = currentIndex;
		for (;;) {
			currentIndex++;
			if (currentIndex >= bytes.Length) { currentIndex = -1; break; }
			if (bytes[currentIndex] >= 0xFE) { break; }
		}
		var valueBytes = currentIndex < 0 ? bytes.Skip(lastIndex+1).ToArray() : bytes.Skip(lastIndex+1).Take(currentIndex-lastIndex-1).ToArray();
		if (valueBytes.Length == 1 && valueBytes[0] == 0xFD) { currentLine.Add(null); }
		else if (valueBytes.Length == 1 && valueBytes[0] == 0xFC) { currentLine.Add(""); }
		else if (valueBytes.Length >= 1) { currentLine.Add(decoder.GetString(valueBytes)); }
		else if ((((currentIndex >= 0 && bytes[currentIndex] == 0xFF) || (currentIndex < 0)) && ((lastIndex < 0) || (lastIndex >= 0 && bytes[lastIndex] == 0xFF))) == false) { throw new Exception("Invalid BSV value byte sequence"); }
		if (currentIndex < 0) { break; }
		else if (currentIndex >= 0 && bytes[currentIndex] == 0xFF) {
			result.Add(currentLine.ToArray());
			currentLine.Clear();
		}
	}
	result.Add(currentLine.ToArray());
	return result.ToArray();
}

// ----------------------------------------------------------------------

static void SaveBsv(string?[][] jaggedArray, string filePath) {
	File.WriteAllBytes(filePath, EncodeBsv(jaggedArray));
}

static string?[][] LoadBsv(string filePath) {
	return DecodeBsv(File.ReadAllBytes(filePath));
}

static void AppendBsv(string?[][] jaggedArray, string filePath) {
	var openFile = () => {
		try { return (File.Open(filePath, FileMode.Open, FileAccess.Write), true); }
		catch (FileNotFoundException exception) { return (File.Open(filePath, FileMode.CreateNew, FileAccess.Write), false); }
	};
	var (handle, existed) = openFile();
	try {
		handle.Position = handle.Length;
		if (existed == true) { handle.Write(new byte[]{0xFF}); }
		handle.Write(EncodeBsv(jaggedArray));
	} finally { handle.Close(); }
}

// ----------------------------------------------------------------------

static string ByteArrayToString(byte[] bytes) {
	return "["+string.Join(", ", bytes)+"]";
}

static string JaggedArrayToString(string?[][] jaggedArray) {
	return "[\n"+String.Join("\n", jaggedArray.Select((line) => "  ["+String.Join(", ", line.Select(x => x == null ? "null" : "\""+x.Replace("\n", "\\n").Replace("\0", "\\0")+"\""))+"]"))+"\n]";
}

static void PrintJaggedArray(string?[][] jaggedArray) {
	Console.WriteLine(JaggedArrayToString(jaggedArray));
}

string?[][] jaggedArray = new []{
	new []{"Hello", "🌎", null, ""},
	new []{"A\0B\nC", "Test 𝄞"}
};

Console.OutputEncoding = System.Text.Encoding.UTF8;

PrintJaggedArray(jaggedArray);
var bytes = EncodeBsv(jaggedArray);
Console.WriteLine(ByteArrayToString(bytes));

var decoded = DecodeBsv(bytes);
PrintJaggedArray(jaggedArray);

SaveBsv(jaggedArray, "Test.bsv");
var loaded = LoadBsv("Test.bsv");
PrintJaggedArray(jaggedArray);

AppendBsv(new []{new []{"ABC"}}, "Append.bsv");

Console.WriteLine("Dev");

File: VB.NET/bsv.vb
' (C) Stefan John / Stenway / Stenway.com / 2023

Imports System
Imports System.Linq
Imports System.Collections.Generic
Imports System.Text
Imports System.IO

Module Bsv
	Function EncodeBsv(jaggedArray As String()()) As Byte()
		Dim parts = New List(Of Byte())()
		Dim lineBreakByte = New Byte() {&Hff}
		Dim valueSeparatorByte = New Byte() {&Hfe}
		Dim nullValueByte = New Byte() {&Hfd}
		Dim emptyStringByte = New Byte() {&Hfc}
		Dim encoder = New UTF8Encoding(False, True)
		Dim isFirstLine = True
		For Each line As String() In jaggedArray
			If isFirstLine = False Then
				parts.Add(lineBreakByte)
			End If
			isFirstLine = False
			Dim isFirstValue = True
			For Each value As String In line
				If isFirstValue = False Then
					parts.Add(valueSeparatorByte)
				End If
				isFirstValue = False
				If value Is Nothing Then
					parts.Add(nullValueByte)
				ElseIf value.Length = 0 Then
					parts.Add(emptyStringByte)
				Else
					parts.Add(encoder.GetBytes(value))
				End If
			Next
		Next
		Dim result = New Byte(parts.Sum(Function(part) part.Length) - 1) {}
		Dim stream = New MemoryStream(result)
		For Each part As Byte() In parts
			stream.Write(part, 0, part.Length)
		Next
		Return result
	End Function
	
	Function DecodeBsv(bytes As Byte()) As String()()
		Dim decoder = New UTF8Encoding(False, True)
		Dim result = New List(Of String())()
		Dim currentLine = New List(Of String)()
		Dim currentIndex = -1
		While True
			Dim lastIndex = currentIndex
			While True
				currentIndex += 1
				If currentIndex >= bytes.Length Then
					currentIndex = -1
					Exit While
				End If
				If bytes(currentIndex) >= &Hfe Then
					Exit While
				End If
			End While
			Dim valueBytes = If(currentIndex < 0, bytes.Skip(lastIndex + 1).ToArray(), bytes.Skip(lastIndex + 1).Take(currentIndex - lastIndex - 1).ToArray())
			If valueBytes.Length = 1 AndAlso valueBytes(0) = &Hfd Then
				currentLine.Add(Nothing)
			ElseIf valueBytes.Length = 1 AndAlso valueBytes(0) = &Hfc Then
				currentLine.Add("")
			ElseIf valueBytes.Length >= 1 Then
				currentLine.Add(decoder.GetString(valueBytes))
			ElseIf (((currentIndex >= 0 AndAlso bytes(currentIndex) = &Hff) OrElse (currentIndex < 0)) AndAlso ((lastIndex < 0) OrElse (lastIndex >= 0 AndAlso bytes(lastIndex) = &Hff))) = False Then
				Throw New Exception("Invalid BSV value byte sequence")
			End If
			If currentIndex < 0 Then
				Exit While
			ElseIf currentIndex >= 0 AndAlso bytes(currentIndex) = &Hff Then
				result.Add(currentLine.ToArray())
				currentLine.Clear()
			End If
		End While
		result.Add(currentLine.ToArray())
		Return result.ToArray()
	End Function
	
	' ----------------------------------------------------------------------

	Sub SaveBsv(jaggedArray As String()(), filePath As String)
		File.WriteAllBytes(filePath, EncodeBsv(jaggedArray))
	End Sub
	
	Function LoadBsv(filePath As String) As String()()
		Return DecodeBsv(File.ReadAllBytes(filePath))
	End Function
	
	Sub AppendBsv(jaggedArray As String()(), filePath As String)
		Dim handle As FileStream
		Dim existed As Boolean
		Try
			handle = File.Open(filePath, FileMode.Open, FileAccess.Write)
			existed = True
		Catch exception As FileNotFoundException
			handle = File.Open(filePath, FileMode.CreateNew, FileAccess.Write)
			existed = False
		End Try
		Try
			handle.Position = handle.Length
			If existed = True Then
				handle.Write(New Byte() {&Hff}, 0, 1)
			End If
			Dim bytes = EncodeBsv(jaggedArray)
			handle.Write(bytes, 0, bytes.Length)
		Finally
			handle.Close()
		End Try
	End Sub
	
	' ----------------------------------------------------------------------

	Function JaggedArrayToString(ByVal jaggedArray As String()()) As String
    	Return "[" & vbLf & String.Join(
			vbLf,
			jaggedArray.Select(Function(line) "  [" & String.Join(
				", ",
				line.Select(Function(x) If(x Is Nothing, "null", """" & x.Replace(vbLf, "\n").Replace(vbNullChar, "\0") & """"))
			) & "]")
		) & vbLf & "]"
	End Function
	
	Sub PrintJaggedArray(ByVal jaggedArray As String()())
		Console.WriteLine(JaggedArrayToString(jaggedArray))
	End Sub
	
	Function ByteArrayToString(bytes As Byte()) As String
		Return "[" & String.Join(", ", bytes) & "]"
	End Function
	
	' ----------------------------------------------------------------------

	Sub Main()
		Console.OutputEncoding = System.Text.Encoding.UTF8
	
		Dim jaggedArray As String()() = {
			New String() {"Hello", "🌎", Nothing, ""},
			New String() {"A" & vbNullChar & "B" & vbLf & "C", "Test 𝄞"}
		}
		PrintJaggedArray(jaggedArray)
		
		Dim bytes = EncodeBsv(jaggedArray)
		Console.WriteLine(ByteArrayToString(bytes))

		Dim decodedJaggedArray = DecodeBsv(bytes)
		PrintJaggedArray(decodedJaggedArray)
		
		SaveBsv(jaggedArray, "Test.bsv")
		
		Dim loadedJaggedArray As String()() = LoadBsv("Test.bsv") 
		PrintJaggedArray(loadedJaggedArray)
		SaveBsv(loadedJaggedArray, "TestResaved.bsv")
		
		AppendBsv({New String() {"ABC"}}, "Append.bsv")
		PrintJaggedArray(LoadBsv("Append.bsv"))
		
		Console.ReadKey()
	End Sub
End Module